lib/beaker/hypervisor/libvirt.rb

```diff
@@ -0,0 +1,152 @@
+# TODO: rework qcow handling to do this:
+# virsh vol-create-as default guest 2048MB
+# virsh vol-upload --pool default guest file.qcow2
+# virsh vol-delete --pool default guest
+module Beaker
+  class Libvirt < Beaker::Hypervisor
+
+    def initialize(hosts, options)
+      require 'tempfile'
+      @options = options
+      @logger = options[:logger]
+      @hosts = hosts
+      @logger.notify "Connecting to libvirt"
+
+    end
+
+    def provision
+      @logger.notify "Provisioning libvirt"
+      #conn = ::Libvirt::open('qemu:///system')
+
+      @hosts.each do |host|
+        file = Tempfile.new(host.name)
+        img = Tempfile.new(host.name + "2")
+        temporary_name = file.path.split('/')[-1]
+        host['tempname'] = temporary_name
+        host['qcowfile'] = img.path
+        @logger.notify "provisioning #{host.name} as #{host['tempname']}"
+        FileUtils.cp host['qcow2'], img.path
+
+        new_dom_xml = <<EOF
+<domain type='kvm'>
+  <name>#{host['tempname']}</name>
+  <memory unit='MiB'>2048</memory>
+  <vcpu>1</vcpu>
+  <os>
+    <type arch='x86_64'>hvm</type>
+    <boot dev='hd'/>
+    <bootmenu enable='no'/>
+  </os>
+  <features>
+    <acpi/>
+    <apic/>
+    <pae/>
+  </features>
+  <clock offset='utc'/>
+  <on_poweroff>destroy</on_poweroff>
+  <on_reboot>restart</on_reboot>
+  <on_crash>restart</on_crash>
+  <devices>
+    <emulator>/usr/bin/qemu-system-x86_64</emulator>
+    <disk type='file' device='disk'>
+      <driver name='qemu' type='qcow2' cache='unsafe'/>
+      <source file='#{img.path}'/>
+      <target dev='sda' bus='sata'/>
+      <address type='drive' controller='0' bus='0' unit='0'/>
+    </disk>
+    <controller type='ide' index='0'>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x1'/>
+    </controller>
+    <input type='mouse' bus='ps2'/>
+    <graphics type='vnc' port='-1' autoport='yes'/>
+    <video>
+      <model type='cirrus' vram='9216' heads='1'/>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
+    </video>
+    <memballoon model='virtio'>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x07' function='0x0'/>
+    </memballoon>
+    <interface type='network'>
+      <source network='default'/>
+      <model type='virtio'/>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
+    </interface>
+    <serial type='pty'>
+      <target port='0'/>
+    </serial>
+    <console type='pty'>
+      <target type='serial' port='0'/>
+    </console>
+  </devices>
+</domain>
+EOF
+        @logger.notify "Creating transient domain ruby-libvirt-tester"
+        file.write(new_dom_xml)
+        file.close
+        `virsh define #{file.path}`
```

Comment:
Best to use something like popen. see https://github.com/puppetlabs/beaker/blob/master/lib/beaker/hypervisor/vagrant.rb#L67 for example.
Perhaps it could be created into a generic helper / util function.